testing-internal/src/main/java/com/linecorp/armeria/internal/testing/AbstractMethodChainingTest.java

```diff
@@ -0,0 +1,93 @@
+import org.reflections.ReflectionUtils;
+import org.reflections.Reflections;
+import org.reflections.adapters.JavaReflectionAdapter;
+import org.reflections.scanners.SubTypesScanner;
+import org.reflections.util.ClasspathHelper;
+import org.reflections.util.ConfigurationBuilder;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Makes sure most builder overrides all overridden methods with the correct return type.
+ */
+public abstract class AbstractMethodChainingTest {
+
+    @Test
+    void methodChaining() {
+        final String packageName = getClass().getPackage().getName();
+        findAllClasses(packageName).stream()
+                                   .map(ReflectionUtils::forName)
+                                   .filter(clazz -> clazz.getSimpleName().endsWith("Builder"))
+                                   .forEach(clazz -> {
+                                       final List<Method> methods = obtainMethods(clazz);
+                                       for (Method m : methods) {
+                                           try {
+                                               final Method overriddenMethod =
+                                                       clazz.getDeclaredMethod(m.getName(),
+                                                                               m.getParameterTypes());
+                                               assertThat(overriddenMethod.getReturnType()).isSameAs(clazz);
+                                           } catch (NoSuchMethodException e) {
+                                               // ignored
+                                           }
+                                       }
+                                   });
+    }
+
+    private static Collection<String> findAllClasses(String packageName) {
+        final ConfigurationBuilder configuration = new ConfigurationBuilder()
+                .filterInputsBy(filePath -> filePath != null && filePath.endsWith(".class"))
+                .setUrls(ClasspathHelper.forPackage(packageName))
+                .setScanners(new SubTypesScanner())
+                .setMetadataAdapter(new JavaReflectionAdapter());
+        final Reflections reflections = new Reflections(configuration);
+        return reflections.getStore().get(SubTypesScanner.class.getSimpleName()).values();
+    }
+
+    private static List<Method> obtainMethods(Class<?> clazz) {
+        final Set<Class<?>> allSuperTypes = ReflectionUtils.getAllSuperTypes(clazz, input -> input != clazz);
+        final ImmutableList<Method> methods = allSuperTypes.stream()
+                                                           .flatMap(sc -> Arrays.stream(sc.getMethods()))
+                                                           .distinct()
+                                                           .collect(toImmutableList());
+        // In general, if parent classes have a build method, did not override the method with that type.
```

Comment:
Could you give an example of this?

Reply:
The things mentioned [here](https://github.com/line/armeria/pull/3220#pullrequestreview-551345764) can be an example.